geode-core/src/integrationTest/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxyTest.java

```diff
@@ -0,0 +1,95 @@
+import static org.assertj.core.api.Assertions.assertThatCode;
+import static org.junit.Assert.assertNull;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.net.InetAddress;
+import java.net.Socket;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+
+import org.junit.Rule;
+import org.junit.Test;
+
+import org.apache.geode.distributed.DistributedMember;
+import org.apache.geode.internal.Version;
+import org.apache.geode.internal.cache.InternalCache;
+import org.apache.geode.internal.net.SocketCloser;
+import org.apache.geode.test.junit.rules.ExecutorServiceRule;
+import org.apache.geode.test.junit.rules.ServerStarterRule;
+
+public class CacheClientProxyTest {
+
+  @Rule
+  public ServerStarterRule serverRule = new ServerStarterRule().withAutoStart();
+
+  @Rule
+  public ExecutorServiceRule executorServiceRule = new ExecutorServiceRule();
+
+  @Test
+  public void closeSocketShouldBeAtomic() {
+
+    CacheServerStats stats = mock(CacheServerStats.class);
+    doNothing().when(stats).incCurrentQueueConnections();
+    InternalCache cache = serverRule.getCache();
+    CacheClientNotifier ccn = mock(CacheClientNotifier.class);
+    SocketCloser sc = mock(SocketCloser.class);
+    when(ccn.getCache()).thenReturn(cache);
+    when(ccn.getAcceptorStats()).thenReturn(stats);
+    when(ccn.getSocketCloser()).thenReturn(sc);
+    Socket socket = mock(Socket.class);
+    InetAddress address = mock(InetAddress.class);
+    when(socket.getInetAddress()).thenReturn(address);
+    when(address.getHostAddress()).thenReturn("localhost");
+    doNothing().when(sc).asyncClose(any(), eq("localhost"), eq(null));
+
+    ClientProxyMembershipID proxyID = mock(ClientProxyMembershipID.class);
+    DistributedMember member = cache.getDistributedSystem().getDistributedMember();
+    when(proxyID.getDistributedMember()).thenReturn(member);
+    CacheClientProxy proxy = new CacheClientProxy(ccn, socket, proxyID, true,
+        Handshake.CONFLATION_DEFAULT, Version.CURRENT, 1L, true,
+        null, null);
+    Future<Void> result1 = executorServiceRule.runAsync(() -> proxy.close());
+    Future<Void> result2 = executorServiceRule.runAsync(() -> proxy.close());
+    Future<Void> result3 = executorServiceRule.runAsync(() -> proxy.close());
+    ((CompletableFuture<Void>) result1).join();
+    ((CompletableFuture<Void>) result2).join();
+    ((CompletableFuture<Void>) result3).join();
+    assertThatCode(() -> result1.get(60, SECONDS)).doesNotThrowAnyException();
+    assertThatCode(() -> result2.get(60, SECONDS)).doesNotThrowAnyException();
+    assertThatCode(() -> result3.get(60, SECONDS)).doesNotThrowAnyException();
```

Comment:
FYI if you capture `CompletableFuture<>` refs above, then these three joins can become one statement:

```
CompletableFuture.allOf(result1,result2,result3).join();
```

Reply:
sounds good. I enhanced the code.

Reply:
nice